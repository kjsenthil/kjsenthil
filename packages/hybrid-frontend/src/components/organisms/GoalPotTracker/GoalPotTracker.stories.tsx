import * as React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { formatCurrency } from '../../../utils/formatters';
import GoalPotTracker, { GoalPotTrackerProps } from './GoalPotTracker';

export default {
  title: 'Digital Hybrid/Organisms/Goal Pot Tracker',
  component: GoalPotTracker,
} as Meta;

const Template: Story<GoalPotTrackerProps> = (args) => <GoalPotTracker {...args} />;

const currencyFormatter = (val: number) =>
  formatCurrency(val, {
    opts: {
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    },
  });

const lumpSumValue = 375000;
const projectionValue = 1500000;
const remainingValue = 100000;
const totalValue = lumpSumValue + projectionValue + remainingValue;

const defaultArgs: GoalPotTrackerProps = {
  title: 'Your retirement pot',
  potTotal: totalValue,
  progressBarProps: {
    currencyFormatter,
    progressBarData: [
      {
        legendProps: { title: 'Lump sum', value: lumpSumValue },
        progress: lumpSumValue / totalValue,
      },
      {
        legendProps: { title: 'From age 67 - 89', value: projectionValue },
        progress: projectionValue / totalValue,
      },
      {
        legendProps: { title: 'Remaining', value: remainingValue },
        progress: remainingValue / totalValue,
      },
    ],
  },
};

export const Default = Template.bind({});
Default.args = defaultArgs;
