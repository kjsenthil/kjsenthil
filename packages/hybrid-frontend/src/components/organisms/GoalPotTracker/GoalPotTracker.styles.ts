import styled from 'styled-components';
import { Theme } from '../../atoms';

// eslint-disable-next-line import/prefer-default-export
export const GoalPotTrackerContainer = styled.div`
  ${({ theme }: { theme: Theme }) => `
    display: flex;
    flex-direction: column;
    gap: ${theme.spacing(2.5)}px;
    width: 100%;
  `}
`;
