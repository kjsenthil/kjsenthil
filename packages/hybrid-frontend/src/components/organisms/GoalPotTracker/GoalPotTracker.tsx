import * as React from 'react';
import { Box, Typography, useTheme } from '../../atoms';
import { ProgressBarWithLegend, ProgressBarWithLegendProps } from '../../molecules';
import { formatCurrency } from '../../../utils/formatters';
import { GoalPotTrackerContainer } from './GoalPotTracker.styles';

export interface GoalPotTrackerProps {
  title: string;
  potTotal: number | undefined;

  progressBarProps: Omit<ProgressBarWithLegendProps, 'progressBarBackgrounds'>;
}

export default function GoalPotTracker({ title, potTotal, progressBarProps }: GoalPotTrackerProps) {
  const theme = useTheme();

  // When the pot total is undefined, there will be no text after the title
  const formattedRetirementTotal =
    potTotal !== undefined
      ? formatCurrency(potTotal, {
          opts: {
            minimumFractionDigits: 0,
            maximumFractionDigits: 0,
          },
        })
      : '';

  let progressBarBackgrounds: string[] | undefined;
  if (progressBarProps.progressBarData.length === 1) {
    progressBarBackgrounds = [];
    if (progressBarProps.progressBarData[0].legendProps.value === undefined) {
      // This happens when there is no income input yet
      progressBarBackgrounds.push(`${theme.palette.tertiary.light2}`);
    } else {
      // This happens when there is only income input
      progressBarBackgrounds.push(theme.palette.tertiary.main);
    }
  }

  return (
    <GoalPotTrackerContainer>
      <Box>
        <Typography component="span" variant="sh1" color="primary" colorShade="dark2">
          {title}
          {formattedRetirementTotal ? ': ' : ''}
        </Typography>
        <Typography component="span" variant="sh1" color="tertiary">{`${
          formattedRetirementTotal ? `${formattedRetirementTotal}` : ''
        }`}</Typography>
      </Box>

      <Box>
        <ProgressBarWithLegend
          progressBarBackgrounds={progressBarBackgrounds}
          {...progressBarProps}
        />
      </Box>
    </GoalPotTrackerContainer>
  );
}
