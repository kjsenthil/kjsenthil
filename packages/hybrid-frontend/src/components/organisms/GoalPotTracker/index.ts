import GoalPotTracker from './GoalPotTracker';

export default GoalPotTracker;
export type { GoalPotTrackerProps } from './GoalPotTracker';
