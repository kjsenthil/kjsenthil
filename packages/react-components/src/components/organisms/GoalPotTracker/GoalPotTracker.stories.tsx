import * as React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import GoalPotTracker, { GoalPotTrackerProps } from './GoalPotTracker';

export default {
  title: 'Digital Hybrid/Organisms/Goal Pot Tracker',
  component: GoalPotTracker,
} as Meta;

const Template: Story<GoalPotTrackerProps> = (args) => <GoalPotTracker {...args} />;

const lumpSumValue = 375000;
const projectionValue = 1500000;
const remainingValue = 100000;
const totalValue = lumpSumValue + projectionValue + remainingValue;

const defaultArgs: GoalPotTrackerProps = {
  title: 'Your retirement pot',
  potTotal: totalValue,
  progressBarProps: {
    progressBarData: [
      {
        legendProps: { title: 'Lump sum', value: lumpSumValue },
        progress: lumpSumValue / totalValue,
      },
      {
        legendProps: { title: 'From age 67 - 89', value: projectionValue },
        progress: projectionValue / totalValue,
      },
      {
        legendProps: { title: 'Remaining', value: remainingValue },
        progress: remainingValue / totalValue,
      },
    ],
  },
};

export const Default = Template.bind({});
Default.args = defaultArgs;
