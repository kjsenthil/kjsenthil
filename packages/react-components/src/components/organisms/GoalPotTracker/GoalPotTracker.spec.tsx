import * as React from 'react';
import { renderWithTheme, screen } from '@tsw/test-util';
import GoalPotTracker from './GoalPotTracker';

const progressBarData = [
  {
    legendProps: { title: 'Title A', value: 10 },
    progress: 0.1,
  },
  {
    legendProps: { title: 'Title B', value: 30 },
    progress: 0.3,
  },
  {
    legendProps: { title: 'Title C', value: 70 },
    progress: 0.7,
  },
];

describe('GoalPotTracker', () => {
  beforeEach(() => {
    renderWithTheme(
      <GoalPotTracker title="Main title" potTotal={100} progressBarProps={{ progressBarData }} />
    );
  });

  it('renders title', () => {
    const title = screen.getByText('Main title:');

    expect(title).toBeInTheDocument();
  });

  it('renders progress bar data', () => {
    const legendTitle1 = screen.getByText('TITLE A');
    const legendTitle2 = screen.getByText('TITLE B');
    const legendTitle3 = screen.getByText('TITLE C');

    const legendValue1 = screen.getByText('£10');
    const legendValue2 = screen.getByText('£30');
    const legendValue3 = screen.getByText('£70');

    expect(legendTitle1).toBeInTheDocument();
    expect(legendTitle2).toBeInTheDocument();
    expect(legendTitle3).toBeInTheDocument();
    expect(legendValue1).toBeInTheDocument();
    expect(legendValue2).toBeInTheDocument();
    expect(legendValue3).toBeInTheDocument();
  });
});
