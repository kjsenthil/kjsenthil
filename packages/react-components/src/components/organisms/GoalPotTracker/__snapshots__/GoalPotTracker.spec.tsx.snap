// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalPotTracker matches snapshot 1`] = `
Object {
  "result": Object {
    "asFragment": [Function],
    "baseElement": .c1 {
  font-size: 1.0625rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: 0.01875rem;
  -moz-letter-spacing: 0.01875rem;
  -ms-letter-spacing: 0.01875rem;
  letter-spacing: 0.01875rem;
  font-weight: 600;
  color: #380c57;
  white-space: break-spaces;
}

.c2 {
  font-size: 1.0625rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: 0.01875rem;
  -moz-letter-spacing: 0.01875rem;
  -ms-letter-spacing: 0.01875rem;
  letter-spacing: 0.01875rem;
  font-weight: 600;
  color: #00c4bd;
  white-space: break-spaces;
}

.c11 {
  font-size: 0.6875rem;
  line-height: 1rem;
  -webkit-letter-spacing: 0.0125rem;
  -moz-letter-spacing: 0.0125rem;
  -ms-letter-spacing: 0.0125rem;
  letter-spacing: 0.0125rem;
  font-weight: 600;
  color: #6a6775;
  white-space: nowrap;
}

.c12 {
  font-size: 0.6875rem;
  line-height: 1rem;
  -webkit-letter-spacing: 0.0125rem;
  -moz-letter-spacing: 0.0125rem;
  -ms-letter-spacing: 0.0125rem;
  letter-spacing: 0.0125rem;
  font-weight: 600;
  color: #380c57;
  white-space: nowrap;
}

.c3 {
  position: relative;
  width: 100%;
  height: 2.375rem;
  border-radius: 6px;
  overflow: hidden;
  background: #fff;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  height: 2.375rem;
  width: 110%;
  border-radius: 6px;
  background: #a5f2ec;
  -webkit-animation: dKzkex 1s ease-out;
  animation: dKzkex 1s ease-out;
  z-index: 1;
}

.c6 {
  position: absolute;
  top: 0;
  left: 0;
  height: 2.375rem;
  width: 40%;
  border-radius: 6px;
  background: linear-gradient(to left,#00e8d9,#00c4bd 50%);
  -webkit-animation: gBYqud 1s ease-out;
  animation: gBYqud 1s ease-out;
  z-index: 1;
}

.c7 {
  position: absolute;
  top: 0;
  left: 0;
  height: 2.375rem;
  width: 10%;
  border-radius: 6px;
  background: #009492;
  -webkit-animation: kiPwco 1s ease-out;
  animation: kiPwco 1s ease-out;
  z-index: 1;
}

.c4 {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  border-radius: 6px;
  border: 2px dashed #ebb300;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 8px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
  overflow: hidden;
}

.c9 {
  min-width: 120px;
  max-width: 50%;
  width: 10%;
  -webkit-animation: eBhLBk 1s ease-out;
  animation: eBhLBk 1s ease-out;
  z-index: 1;
}

.c13 {
  min-width: 120px;
  max-width: 50%;
  width: 30%;
  -webkit-animation: gyLRrW 1s ease-out;
  animation: gyLRrW 1s ease-out;
  z-index: 1;
}

.c14 {
  min-width: 120px;
  max-width: 50%;
  width: 70%;
  -webkit-animation: fhnplK 1s ease-out;
  animation: fhnplK 1s ease-out;
  z-index: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 20px;
  width: 100%;
}

<body>
      <div>
        <div
          class="c0"
        >
          <div
            class="MuiBox-root MuiBox-root-1"
          >
            <span
              class="MuiTypography-root c1 MuiTypography-body1"
            >
              Main title
              : 
            </span>
            <span
              class="MuiTypography-root c2 MuiTypography-body1"
            >
              £100
            </span>
          </div>
          <div
            class="MuiBox-root MuiBox-root-2"
          >
            <div
              aria-valuenow="110.00000000000001"
              class="c3"
              role="progressbar"
            >
              <div
                class="c4"
              />
              <div
                class="c5"
              />
              <div
                class="c6"
              />
              <div
                class="c7"
              />
            </div>
            <div
              class="MuiBox-root MuiBox-root-3"
              data-testid="Spacer"
            />
            <div
              class="c8"
            >
              <div
                class="c9"
                width="10"
              >
                <div
                  class="c10"
                >
                  <div>
                    <h6
                      class="MuiTypography-root c11 MuiTypography-body1"
                    >
                      TITLE A
                    </h6>
                    <div
                      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-item"
                      >
                        <h6
                          class="MuiTypography-root c12 MuiTypography-body1"
                        >
                          £10
                        </h6>
                      </div>
                      <div
                        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="c13"
                width="30"
              >
                <div
                  class="c10"
                >
                  <div>
                    <h6
                      class="MuiTypography-root c11 MuiTypography-body1"
                    >
                      TITLE B
                    </h6>
                    <div
                      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-item"
                      >
                        <h6
                          class="MuiTypography-root c12 MuiTypography-body1"
                        >
                          £30
                        </h6>
                      </div>
                      <div
                        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="c14"
                width="70"
              >
                <div
                  class="c10"
                >
                  <div>
                    <h6
                      class="MuiTypography-root c11 MuiTypography-body1"
                    >
                      TITLE C
                    </h6>
                    <div
                      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-item"
                      >
                        <h6
                          class="MuiTypography-root c12 MuiTypography-body1"
                        >
                          £70
                        </h6>
                      </div>
                      <div
                        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </body>,
    "container": <div>
      <div
        class="sc-jHcXXw skTpp"
      >
        <div
          class="MuiBox-root MuiBox-root-1"
        >
          <span
            class="MuiTypography-root sc-bdnxRM eryqOb MuiTypography-body1"
          >
            Main title
            : 
          </span>
          <span
            class="MuiTypography-root sc-bdnxRM fHGmgY MuiTypography-body1"
          >
            £100
          </span>
        </div>
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <div
            aria-valuenow="110.00000000000001"
            class="sc-fujyAs iugzGY"
            role="progressbar"
          >
            <div
              class="sc-jrsJWt iuJOlM"
            />
            <div
              class="sc-pNWdM gajYlp"
            />
            <div
              class="sc-pNWdM dppDyt"
            />
            <div
              class="sc-pNWdM ccIwFe"
            />
          </div>
          <div
            class="MuiBox-root MuiBox-root-3"
            data-testid="Spacer"
          />
          <div
            class="sc-jUfyBS dPBKsx"
          >
            <div
              class="sc-jQAxuV dKvpIv"
              width="10"
            >
              <div
                class="sc-gSYDnn cftLSt"
              >
                <div>
                  <h6
                    class="MuiTypography-root sc-bdnxRM crCxET MuiTypography-body1"
                  >
                    TITLE A
                  </h6>
                  <div
                    class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item"
                    >
                      <h6
                        class="MuiTypography-root sc-bdnxRM jXNAHh MuiTypography-body1"
                      >
                        £10
                      </h6>
                    </div>
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-jQAxuV bgGgSq"
              width="30"
            >
              <div
                class="sc-gSYDnn cftLSt"
              >
                <div>
                  <h6
                    class="MuiTypography-root sc-bdnxRM crCxET MuiTypography-body1"
                  >
                    TITLE B
                  </h6>
                  <div
                    class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item"
                    >
                      <h6
                        class="MuiTypography-root sc-bdnxRM jXNAHh MuiTypography-body1"
                      >
                        £30
                      </h6>
                    </div>
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-jQAxuV gHXCAo"
              width="70"
            >
              <div
                class="sc-gSYDnn cftLSt"
              >
                <div>
                  <h6
                    class="MuiTypography-root sc-bdnxRM crCxET MuiTypography-body1"
                  >
                    TITLE C
                  </h6>
                  <div
                    class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item"
                    >
                      <h6
                        class="MuiTypography-root sc-bdnxRM jXNAHh MuiTypography-body1"
                      >
                        £70
                      </h6>
                    </div>
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>,
    "debug": [Function],
    "findAllByAltText": [Function],
    "findAllByDisplayValue": [Function],
    "findAllByLabelText": [Function],
    "findAllByPlaceholderText": [Function],
    "findAllByRole": [Function],
    "findAllByTestId": [Function],
    "findAllByText": [Function],
    "findAllByTitle": [Function],
    "findByAltText": [Function],
    "findByDisplayValue": [Function],
    "findByLabelText": [Function],
    "findByPlaceholderText": [Function],
    "findByRole": [Function],
    "findByTestId": [Function],
    "findByText": [Function],
    "findByTitle": [Function],
    "getAllByAltText": [Function],
    "getAllByDisplayValue": [Function],
    "getAllByLabelText": [Function],
    "getAllByPlaceholderText": [Function],
    "getAllByRole": [Function],
    "getAllByTestId": [Function],
    "getAllByText": [Function],
    "getAllByTitle": [Function],
    "getByAltText": [Function],
    "getByDisplayValue": [Function],
    "getByLabelText": [Function],
    "getByPlaceholderText": [Function],
    "getByRole": [Function],
    "getByTestId": [Function],
    "getByText": [Function],
    "getByTitle": [Function],
    "queryAllByAltText": [Function],
    "queryAllByDisplayValue": [Function],
    "queryAllByLabelText": [Function],
    "queryAllByPlaceholderText": [Function],
    "queryAllByRole": [Function],
    "queryAllByTestId": [Function],
    "queryAllByText": [Function],
    "queryAllByTitle": [Function],
    "queryByAltText": [Function],
    "queryByDisplayValue": [Function],
    "queryByLabelText": [Function],
    "queryByPlaceholderText": [Function],
    "queryByRole": [Function],
    "queryByTestId": [Function],
    "queryByText": [Function],
    "queryByTitle": [Function],
    "rerender": [Function],
    "unmount": [Function],
  },
}
`;
